components/entity-access-rules/api/src/main/java/org/phenotips/data/permissions/internal/DefaultPermissionsManager.java

```diff
@@ -0,0 +1,157 @@
+ * along with this program.  If not, see http://www.gnu.org/licenses/
+ */
+package org.phenotips.data.permissions.internal;
+
+import org.phenotips.data.Patient;
+import org.phenotips.data.permissions.AccessLevel;
+import org.phenotips.data.permissions.EntityAccess;
+import org.phenotips.data.permissions.PermissionsManager;
+import org.phenotips.data.permissions.Visibility;
+import org.phenotips.data.permissions.events.EntityRightsUpdatedEvent;
+import org.phenotips.data.permissions.internal.access.AccessHelper;
+import org.phenotips.data.permissions.internal.visibility.VisibilityHelper;
+import org.phenotips.entities.PrimaryEntity;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.observation.ObservationManager;
+
+import java.util.Collection;
+import java.util.Iterator;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * Default implementation of the {@link PermissionsManager}.
+ *
+ * @version $Id$
+ * @since 1.0M9, updated 1.4
+ */
+@Component
+@Singleton
+public class DefaultPermissionsManager implements PermissionsManager
+{
+    @Inject
+    private ObservationManager observationManager;
+
+    @Inject
+    private PermissionsHelper permissionsHelper;
+
+    @Inject
+    private AccessHelper accessHelper;
+
+    @Inject
+    private VisibilityHelper visibilityHelper;
+
+    @Override
+    @Nonnull
+    public Collection<Visibility> listVisibilityOptions()
+    {
+        return this.visibilityHelper.listVisibilityOptions();
+    }
+
+    @Override
+    @Nonnull
+    public Collection<Visibility> listAllVisibilityOptions()
+    {
+        return this.visibilityHelper.listAllVisibilityOptions();
+    }
+
+    @Override
+    @Nonnull
+    public Visibility getDefaultVisibility()
+    {
+        return this.visibilityHelper.getDefaultVisibility();
+    }
+
+    @Override
+    @Nonnull
+    public Visibility resolveVisibility(@Nullable final String name)
+    {
+        return this.visibilityHelper.resolveVisibility(name);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public Collection<Patient> filterByVisibility(Collection<Patient> patients, Visibility requiredVisibility)
+    {
+        return (Collection<Patient>) (Collection<?>) filterVisible((Collection<PrimaryEntity>) (Collection<?>) patients,
+            requiredVisibility);
+    }
+
+    @Override
+    @Nonnull
+    public Collection<PrimaryEntity> filterVisible(
+        @Nullable final Collection<PrimaryEntity> entities,
+        @Nullable final Visibility requiredVisibility)
+    {
+        return this.visibilityHelper.filterByVisibility(entities, requiredVisibility);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public Iterator<Patient> filterByVisibility(Iterator<Patient> patients, Visibility requiredVisibility)
+    {
+        return (Iterator<Patient>) (Iterator<?>) filterVisible((Iterator<PrimaryEntity>) (Iterator<?>) patients,
```

Comment:
I'm wondering if there is a nicer way to cast thing around, e.g. declare `filterVisible()` as accepting `Iterator<? extends PrimaryEntity>` or something else (could not find anything but `? extends`, but maybe new Java 8 ways were still deep down in the list of results)